### PR TITLE
feat: Support mongodb transactions for import application flow

### PR DESCRIPTION
**Pull request by abhvsn**
## Description

Import application flow is a multi-stage process where we touch all the mongo DB collections required to store the application in DB, which today includes pages, actions, JSObjects, etc. When the flow is complete then only we can say the application is successfully imported and DB won't have any stale objects stored. But in a negative scenario where the flow might break because of some unknown exceptions, objects stored earlier become stale and reside in DB forever. With this PR we are establishing the infra for transactions to avoid saving stale DB objects. This is achieved by a rollback mechanism in case an exception is thrown in the middle of execution.  

Note: Since transactions are built on concepts of logical sessions they require mecahnics (like oplog) which are only available in replica set environment.

You can always convert a standalone to a single noded replica set and transactions will work with this one node.

[https://docs.mongodb.com/manual/tutorial/convert-standalone-to-replica-set/](https://docs.mongodb.com/manual/tutorial/convert-standalone-to-replica-set/?_ga=2.53712296.1453502529.1669195955-2065030687.1664009965)

```
For local dev earlier we used to run the single node without replica set. Henceforth if we want to take the advantage of transactions please follow the steps to run mongod single node in local with replica set:

- Upgrade the MongoDB version to 4.4 or higher if it’s 4.2 or lower (https://www.mongodb.com/docs/manual/release-notes/4.4/#std-label-4.4-upgrade)
- Close the mongoDB instance running in your local
- Start the mongoDB in replica set mode and initiate the replica set
    - mongod --port 27017 --dbpath <path/to/db> --replSet <replica-set-name> && mongo --eval “rs.initiate()”
- One can use following commands to check replica set status: 
    - mongo appsmith
    - rs.status()
- By this time you should have the mongo running with replica set
```
Screenshot 2022-07-01 at 10 31 27 PM

This also means mongodb connection string used in env file will now include the replica-set name if one wants to leverage transactions:
`mongodb://localhost:27017/appsmith?replicaSet={replica-set-name}`

Fixes https://github.com/appsmithorg/appsmith/issues/14543

## Type of change

- Bug fix (non-breaking change which fixes an issue)
- This change requires a documentation update

## How Has This Been Tested?

> JUnit 
> Manual test

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: app/server/appsmith-server/src/test/java/com/appsmith/server/configurations/EmbeddedMongoConfig.java [hunk 1]
```diff
@@ -1,21 +1,93 @@
...
+     */
     @Bean
-    public ImmutableMongodConfig prepareMongodConfig() throws IOException {
-        ImmutableMongodConfig mongoConfigConfig = MongodConfig.builder()
-                .version(Version.Main.V4_2)
-                .build();
-        return mongoConfigConfig;
+    @ConditionalOnMissingBean
+    public MongodConfig embeddedMongoConfiguration(EmbeddedMongoProperties embeddedProperties) throws IOException {
+        ImmutableMongodConfig.Builder builder = MongodConfig.builder().version(Version.Main.V4_4);
```
Comment (abhvsn) [inline app/server/appsmith-server/src/test/java/com/appsmith/server/configurations/EmbeddedMongoConfig.java:63]: We need to upgrade to v4.4 on the prod instance as well to support the transaction.

File: app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/ce/ImportExportApplicationServiceCEImpl.java [hunk 2]
```diff
@@ -1148,7 +1150,8 @@ private Mono<Application> importApplicationInWorkspace(String workspaceId,
                                 analyticsService.sendEvent(AnalyticsEvents.UNIT_EXECUTION_TIME.getEventName(), tuple.getT2().getUsername(), data);
                                 return application;
                             });
-                });
+                })
+                .as(transactionalOperator::transactional);
```
Comment (abhvsn) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/ce/ImportExportApplicationServiceCEImpl.java:1154]: This line will enable the transactions in the import application flow.

**Comment by sharat87** [comment]
@abhvsn, prod DB is at v4.4 now. :)

**Comment by AnaghHegde** [comment]
@sharat87 Did we not upgrade it to 4.6x in the scheduled downtime of v1.8.0?

**Comment by AnaghHegde** [comment]
@sharat87 When are we planning to upgrade the mongo version to 5x?

**Comment by AnaghHegde** [comment]
@sharat87 Since Abhijeet is off, I will be picking this up. Can you please review this PR when you have time?

**Review by sidhantgoel — COMMENTED** [review]
LGTM, jut 1 thing, should we a positive test case also, or it will be taken care in other testcases?

**Comment by AnaghHegde** [comment]
It will be handled by existing scenarios.

**Review by sidhantgoel — APPROVED** [review]

**Pull request merged**